Лабораторная работа №5. Рефакторинг

TemZZZ opened the pull request.

Commit:
feature(View): добавлен и реализован класс RadioComponentControl

- класс элемента отбражения информации об уже созданном радиокомпоненте или предоставляющий информацию для создания нового радиокомпонента
- реализовано свойство ReadOnly
- реализовано свойство RadioComponent, позволяющее получить информацию о радиокомпоненте или создать новый

Commit:
style(View.RadioComponentControl): изменен внешний вид компонента

Commit:
refactor(View.AddRadioComponentForm): класс переписан с использованием RadioComponentControl

Commit:
feature(View.RadioComponentControl): добавлен и реализован метод SetDefaultState

- метод переводит элемент в состояние по умолчанию (не выделена ни одна радиокнопка, нет физической величины и единиц измерния)
- если свойству RadioComponent присвоить null, то элемент переводится в состояние по умолчанию
- элемент инициализирется состоянием по умолчанию

Commit:
fix(View.AddRaddioComponentForm.AddRadioComponentButton_Click): добавлена проверка аргумента на null

Commit:
feature(View.MainForm): добавлен элемент отбражения информации о радиокомпоненте

- изменена структура формы: все кнопки и поля управления перенесены в правую часть
- информация о выбранном радиокомпоненте дублируется в элементе отображения информации

Commit:
style(View.RadioComponentControl): в методе set свойства RadioComponent удалены ненужные переменные quantity и unit

Commit:
feature(View.MainForm): реализовано изменение выбранного радиокомпонента

- добавлена кнопка Изменить
- добавлен и реализован метод ModifyButton_Click

Commit:
style(View.MainForm): изменена форма

- установлен минимальный размер формы
- назначены anchor'ы элементам

iKalentii commented on View/RadioComponentControl.cs:
```diff
@@ -0,0 +1,158 @@
+
+			set
+			{
+				if ((value is null) || !(value is RadioComponentBase))
```
Проверка на RadioComponentBase - лишняя

iKalentii commented on View/RadioComponentControl.cs:
```diff
@@ -0,0 +1,158 @@
+			{
+				var radioComponentValue = valueDoubleTextBox.GetValue();
+
+				if (resistorRadioButton.Checked)
```
Собрал бы все радиоэлементы в коллекцию и работал бы с ними как с коллекцией

Commit:
refactor(Model.RadioComponentBase): проверка double на корректность вынесена в метод CheckNamedDoubleParameter

- вынесена из сеттера Value
- вынесена из GetImpedance

Commit:
style(Model.Resistor): свойства Unit, Type, Quantity переписаны без get

Commit:
style(Model.Inductor): свойства Unit, Type, Quantity переписаны без get

Commit:
style(Model.Capacitor): свойства Unit, Type, Quantity переписаны без get

Commit:
refactor(View.MainForm): изменен метод SetupRadioComponentsDataGridView

-  вместо списка элементов структуры NameHeaderTextPair используется список элементов именованых кортежей (string columnName, string headerText)

Commit:
refactor(View.MainForm): удалена структура NameHeaderTextPair

Commit:
style(Model.RadioComponentBase.CheckNamedDoubleParameter): изменен комментарий

Commit:
style(Model.RadioComponentBase): изменены комментарии

- комментарий метода CheckNamedDoubleParameter дополнен выбрасываемыми исключениями
- из комментария свойства Value удалены выбрасываемые исключения
- из комментария методв GetImpedance удалены выбрасываемые исключения

Commit:
style(ConsoleLoader.Main): переименована константа exitCharacter в _exitCharacter и добавлен модификатор доступа private к ней

Commit:
style(View.AddRadioComponentForm): изменены комментарии

- изменен комментарий к методу GenerateRandomDataButton_Click
- изменен комментарий к методу AddRadioComponentButton_Click

Commit:
style(View.SearchRadioComponentForm): добавлены модификаторы доступа private к константам

style(View.SearchRadioComponentForm.GetRadioComponentTypeComboBoxItems): изменен комментарий

Commit:
style(View.SetRadioComponentLoadOptionForm.OkButton_Click): изменен комментарий

Commit:
style(View.SetRadioComponentSaveOptionForm.OkButton_Click): изменен комментарий

Commit:
feature(View.RadioComponentControl): в сеттере свойства RadioComponent удалена проверка на RadioComponentBase

Commit:
refactor(View.RadioComponentControl): изменена работа с радиокнопками

- добавлена инициализация списка радиокнопок _radioButtons в конструктор
- добавление обработчика события CheckedChanged радиокнопок RadioButton_CheckedChanged добавляется в цикле в конструкторе
- в методе SetDefaultState состояние Checked всех радиокнопок изменяется в цикле
- в сеттере свойства ReadOnly состояние Enabled всех радиокнопок изменяется в цикле
- в методе RadioButton_CheckedChanged множество if-ов переписано в виде цикла благодаря списку пар значений radioButtonToQuantityUnitTextMap
- в сеттере свойства RadioComponent множество if-ов переписано в виде цикла благодаря списку пар значений typeToRadioButtonMap

Commit:
feature(Model): добавлено перечисление типов радиокомпонентов

Commit:
feature(Model): добавлен и реализован класс RadioComponentFactory

- реализован метод CreateRadioComponent, возвращающий радиокомпонент нужного типа

Commit:
refactor(View.RadioComponentControl): геттер свойства RadioComponent переписан с использованием фабрики RadioComponentFactory

Commit:
refactor(View.AddRadioComponentForm.GenerateRandomDataButton_Click): переписан метод

- радиокомпонент создается фабрикой RadioComponentFactory
- делитель divisor выбирается по номеру радиокнопки, случайно генерируему генератором randomIntGenerator

Commit:
refactor(View.RadioComponentControl): изменены методы

- в сеттере RadioComponent теперь используется словарь вместо списка
- в методе RadioButton_CheckedChanged теперь используется словарь вместо списка

Commit:
feature(View.RadioComponentFactory): удален перечислитель TypeToRadioComponentMap

feature(View.RadioComponentFactory): добавлен и реализован метод GetTypeToRadioComponentMap

feature(View.RadioComponentFactory): метод CreateRadioComponent переписан через возврат радиокомпонента по ключу из словаря GetTypeToRadioComponentMap

Commit:
style(View.RadioComponentFactory.GetTypeToRadioComponentMap): изменен комментарий

feature(View.RadioComponentFactory.CreateRadioComponent): добавлена проверка входных параметров

- проверяется значение radioComponentType, чтоб значение было не больше количества элементов в словаре GetTypeToRadioComponentMap

Commit:
feature(Model.UnitTests): добавлен тестовый класс RadioComponentFactoryTests

- добавлен и реализован метод CreateRadioComponent_AssignedGoodValues_ReturnsObjectOfExpectedType
- добавлены тесты для тестирования метода CreateRadioComponent (атрибуты перед методом CreateRadioComponent_AssignedGoodValues_ReturnsObjectOfExpectedType, тесты пока без имен)

iKalentii commented on Model/RadioComponentFactory.cs:
```diff
@@ -0,0 +1,61 @@
+			}
+			else
+			{
+				return null;
```
Тут может быть есть смысл прям упасть (кинуть исключение), чтобы показать, что пришло какое-то странное значение. Но обращение в словарь по отсутсвующему ключу к тому же должно было привести.

iKalentii commented on View/Forms/AddRadioComponentForm.cs:
```diff
@@ -86,83 +62,52 @@ private void CancelButton_Click(object sender, EventArgs e)
         {
-            var randomIntGenerator = new Random();
-
             const int maxRadioButtonNumber = 3;
```
Можно было вместо константы брать count от словаря.

iKalentii commented on View/RadioComponentControl.cs:
```diff
@@ -0,0 +1,154 @@
+					= string.Join(", ", value.Quantity, value.Unit);
+
+				var typeToRadioButtonMap
+					= new Dictionary<Type, RadioButton>
```
Собрал бы эту информацию также в общий словарь

iKalentii commented on View/RadioComponentControl.cs:
```diff
@@ -0,0 +1,154 @@
+			var radioButtonToQuantityUnitTextMap
+				= new Dictionary<RadioButton, string>
+				{
+					[resistorRadioButton] = "Сопротивление, Ом",
```
И эту информацию

iKalentii commented on View/RadioComponentControl.cs:
```diff
@@ -0,0 +1,154 @@
+				var typeToRadioButtonMap
+					= new Dictionary<Type, RadioButton>
+					{
+						[typeof(Resistor)] = resistorRadioButton,
```
Типы лучше перенести в фабрику, чтобы в коде приложения не было никаких ссылок на конкретные типы, а была информация только о фабрике.

Commit:
refactor(View.AddRadioComponentForm.GenerateRandomDataButton_Click): удалена константа maxRadioButtonNumber

- вместо maxRadioButtonNumber используется radioButtonNumberToDivisorMap.Count

Commit:
feature(Model.RadioComponentFactory.CreateRadioComponent): удалена проверка (int)radioComponentType <= typeToRadioComponentMap.Count

- в случае передачи некорректного значения radioComponentType теперь будет выбрасываться исключение KeyNotFoundException
- комментарий к методу дополнен возможностью выбрасывания исключения

Commit:
style(RadioComponentsCollection): все пробельные отступы в началах строк заменены на табуляции

Commit:
refactor(View.MainForm.SetupRadioComponentsDataGridView): properties переименован в columnNameToHeaderTextMap

- в цикле используется именованый кортеж для укорочения инструкций

Commit:
feature(Model.RadioComponentFactory): добавлен и реализован метод GetRadioComponentType

- добавлен словарь _typeToRadioComponentTypeMap
- метод GetRadioComponentType возвращает соответствующий тип RadioComponentType типу Type из словаря _typeToRadioComponentTypeMap

Commit:
refactor(View.RadioComponentControl): вместо _radioButtons используется _radioButtonInfoDictionary

- удален список _radioButtons
- создан список List<(RadioButton radioButton, RadioComponentType type, string quantityUnitText)> _radioButtonInfoDictionary, инициализируется в конструкторе
- в конструкторе обработчики RadioButton_CheckedChanged добавляются в цикле с ипользованием _radioButtonInfoDictionary
- сеттер свойства ReadOnly переписан с использованием _radioButtonInfoDictionary
- геттер свойства RadioComponent переписан через использование _radioButtonInfoDictionary
- в сеттере свойства RadioComponent удален словарь typeToRadioButtonMap, тип радиокомпонента возвращается фабрикой radioComponentfactory, цикл переписан с использованием _radioButtonInfoDictionary
- цикл метода SetDefaultState переписан с использованием _radioButtonInfoDictionary
- исправлен комментарий к методу RadioButton_CheckedChanged
- в методе RadioButton_CheckedChanged удален словарь radioButtonToQuantityUnitTextMap, цикл переписан с использованием _radioButtonInfoDictionary

Commit:
feature(View.MainForm.ModifyButton_Click): добавлена на null проверка радиокомпонента от modifyRadioComponentControl

Commit:
refactor(View.AddRadioComponentForm.AddRadioComponentButton_Click): инвертирован оператор if

Commit:
refactor(Model.UnitTests.RadioComponentFactoryTests.CreateRadioComponent_PassedGoodValues_ReturnsObjectOfExpectedType): переименован из CreateRadioComponent_AssignedGoodValues_ReturnsObjectOfExpectedType

Commit:
Revert "refactor(Model.UnitTests.RadioComponentFactoryTests.CreateRadioComponent_PassedGoodValues_ReturnsObjectOfExpectedType): переименован из CreateRadioComponent_AssignedGoodValues_ReturnsObjectOfExpectedType"

This reverts commit 5afade4833808f6fe94ef22c10e5abe165a3e068.

Commit:
refactor(Model.UnitTests.RadioComponentFactoryTests.CreateRadioComponent_ReceivedGoodValues_ReturnsObjectOfExpectedType): переименован из CreateRadioComponent_AssignedGoodValues_ReturnsObjectOfExpectedType

Commit:
refactor(Model.UnitTests.RadioComponentFactoryTests): изменен класс

- добавлен тестовый метод GetRadioComponentType_ReceivedGoodValue_ReturnsValue
- удалены три атрибута TestCase
- добавлен список _radioComponentInfoDictionary, ставящий в соответствие типы и объекты
- добавлены формирователи тестовых случаев CreateRadioComponentTestCases и GetRadioComponentTypeTestCases
- из сигнатуры метода CreateRadioComponent_ReceivedGoodValues_ReturnsObjectOfExpectedType удален параметр radioComponentValue, выполняющий ту же функцию параметр создается в теле метода
- добавлены тесты методов CreateRadioComponent (добавлен атрибут TestCaseSource перед методом CreateRadioComponent_ReceivedGoodValues_ReturnsObjectOfExpectedType) и GetRadioComponentType (добавлен атрибут TestCaseSource перед иетодом GetRadioComponentType_ReceivedGoodValue_ReturnsValue)

Commit:
refactor(View.RadioComponentControl): добавлен и реализован метод SetupRadioButtons

- в метод вынесены из конструктора инструкции инициализации списка _radioButtonInfoDictionary и назначения радиокнопкам обработчиков событий
- в конструкторе вызывается этот метод после InitializeComponent

Commit:
feature(Model.RadioComponentFactory): добавлен и реализован метод CreateRandomRadioComponent

- метод возвращает случайно сгенерированный радиокомпонент

Commit:
refactor(View.AddRadioComponentForm.GenerateRandomDataButton_Click): метод переписан через CreateRandomRadioComponent от фабрики

Commit:
feature(Model.RadioComponentFactory): сделана статической

Commit:
refactor(View.RadioComponentControl): изменено обращение к фабрике

- в свойстве RadioComponent вызывается метод CreateRadioComponent статического класса RadioComponentFactory

Commit:
refactor(View.AddRadioComponentForm.GenerateRandomDataButton_Click): изменено обращение к фабрике

- вызывается метод CreateRandomRadioComponent статического класса RadioComponentFactory

Commit:
style(View.AddRadioComponentForm): удалена неиспользуемая директива using
